File: libraries/node/src/parser/assembly.ts
import { UrcError } from "../error.js";
import type { Chart, Judgment, Layout, Note, NoteType } from "../model.js";
import { NOTE_TOKENS, REQUIRED_SECTIONS, rule } from "../strings.js";
import type { ParseState } from "./state.js";

const NOTE_TYPES: Readonly<Record<string, NoteType>> = Object.fromEntries(
	NOTE_TOKENS.map(token => [token, token])
);

/**
 * Validates notes (Phase C) and assembles the final Chart.
 */
export function build(state: ParseState, endLine: number): Chart {
	for (const name of REQUIRED_SECTIONS) 
		if (!state.seen.has(name)) 
			throw new UrcError(rule(2), endLine, `missing required section: ${name}`);
	
	const layoutPair = state.layoutType;
	if (layoutPair === null) 
		throw new Error("internal error: Layout finalized without Type");
	
	const [keys, specialKeys] = layoutPair;
	const layout: Layout = { keys, specialKeys, specialLanes: state.special };
	const total = keys + specialKeys;

	const ordered = [...state.notes].sort((a, b) => a.timestamp - b.timestamp || a.lane - b.lane);
	const notes: Note[] = [];
	const openLs = new Map<number, number>();
	for (const raw of ordered) {
		if (raw.timestamp < 0) 
			throw new UrcError(rule(22), raw.line, "note timestamps must be non-negative");
		
		if (raw.lane < 0 || raw.lane >= total) 
			throw new UrcError(rule(18), raw.line, `lane out of range: ${raw.lane}`);
		
		const noteType = NOTE_TYPES[raw.typeToken];
		if (noteType === undefined) 
			throw new UrcError(rule(19), raw.line, `unknown note type: '${raw.typeToken}'`);
		
		if (noteType === "LE") {
			if (!openLs.has(raw.lane)) 
				throw new UrcError(rule(20), raw.line, `LE without an open LS on lane ${raw.lane}`);
			
			openLs.delete(raw.lane);
		} else if (noteType === "LS") {
			if (openLs.has(raw.lane)) 
				throw new UrcError(rule(21), raw.line, `overlapping long notes on lane ${raw.lane}`);
			
			openLs.set(raw.lane, raw.line);
		}
		notes.push({ timestampMs: raw.timestamp, lane: raw.lane, type: noteType });
	}
	const firstOpen = openLs.entries().next().value;
	if (firstOpen !== undefined) {
		const [lane, line] = firstOpen;
		throw new UrcError(rule(20), line, `unterminated LS on lane ${lane}`);
	}

	let judgment: Judgment | null = null;
	if (state.windows !== null && state.rates !== null) 
		judgment = { windows: state.windows, rates: state.rates };
	
	const metadataValue = (name: string): string => {
		const value = state.metadata.get(name);
		if (value === undefined) 
			throw new Error(`internal error: metadata field ${name} is validated`);
		
		return value;
	};
	return {
		formatVersion: state.version,
		metadata: {
			original: metadataValue("Original"),
			title: metadataValue("Title"),
			artist: metadataValue("Artist"),
			creator: metadataValue("Creator"),
			version: metadataValue("Version")
		},
		judgment,
		layout,
		timing: state.timing,
		notes
	};
}

File: libraries/rust/src/parser/assembly.rs
//! Chart assembly: note validation (Phase C) and model construction.

use std::collections::BTreeMap;

use super::state::ParseState;
use crate::error::{Result, UrcError};
use crate::model::{Chart, Judgment, Layout, Metadata, Note, NoteType};
use crate::strings::{REQUIRED_SECTIONS, rule};

pub(super) fn build(state: ParseState, end_line: u32) -> Result<Chart> {
    for name in REQUIRED_SECTIONS {
        let index = crate::strings::section_index(name).unwrap();
        if !state.seen[index] {
            return Err(UrcError::new(
                rule(2),
                end_line,
                format!("missing required section: {name}"),
            ));
        }
    }
    let ParseState {
        version,
        metadata,
        windows,
        rates,
        layout_type,
        special,
        timing,
        mut notes,
        ..
    } = state;
    let (keys, special_keys) = layout_type.expect("Layout is required and finalized");
    let special_lanes = special.map(|lanes| {
        lanes
            .iter()
            .map(|&lane| u32::try_from(lane).expect("special lanes are validated"))
            .collect::<Vec<u32>>()
    });
    let layout = Layout {
        keys,
        special_keys,
        special_lanes,
    };
    let total = layout.total_lanes();

    notes.sort_by_key(|note| (note.timestamp, note.lane));
    let mut chart_notes = Vec::with_capacity(notes.len());
    let mut open_ls: BTreeMap<i64, u32> = BTreeMap::new();
    for raw in &notes {
        if raw.timestamp < 0 {
            return Err(UrcError::new(
                rule(22),
                raw.line,
                "note timestamps must be non-negative",
            ));
        }
        if raw.lane < 0 || raw.lane as u64 >= total {
            return Err(UrcError::new(
                rule(18),
                raw.line,
                format!("lane out of range: {}", raw.lane),
            ));
        }
        let note_type = match raw.type_token.as_str() {
            "N" => NoteType::N,
            "LS" => NoteType::Ls,
            "LE" => NoteType::Le,
            "M" => NoteType::M,
            "F" => NoteType::F,
            other => {
                return Err(UrcError::new(
                    rule(19),
                    raw.line,
                    format!("unknown note type: {other:?}"),
                ));
            }
        };
        match note_type {
            NoteType::Le => match open_ls.remove(&raw.lane) {
                None => {
                    return Err(UrcError::new(
                        rule(20),
                        raw.line,
                        format!("LE without an open LS on lane {}", raw.lane),
                    ));
                }
                Some(_) => {}
            },
            NoteType::Ls => {
                if open_ls.contains_key(&raw.lane) {
                    return Err(UrcError::new(
                        rule(21),
                        raw.line,
                        format!("overlapping long notes on lane {}", raw.lane),
                    ));
                }
                open_ls.insert(raw.lane, raw.line);
            }
            _ => {}
        }
        chart_notes.push(Note {
            timestamp_ms: raw.timestamp,
            lane: raw.lane as u32,
            note_type,
        });
    }
    if let Some((&lane, &line)) = open_ls.iter().next() {
        return Err(UrcError::new(
            rule(20),
            line,
            format!("unterminated LS on lane {lane}"),
        ));
    }

    let field = |name: &str| -> String {
        metadata
            .iter()
            .find(|(known, _)| known == name)
            .map(|(_, value)| value.clone())
            .unwrap_or_else(|| panic!("metadata field {name} is validated"))
    };
    let judgment = windows.map(|windows| Judgment {
        windows,
        rates: rates.expect("Judgment fields are paired"),
    });
    Ok(Chart {
        format_version: version,
        metadata: Metadata {
            original: field("Original"),
            title: field("Title"),
            artist: field("Artist"),
            creator: field("Creator"),
            version: field("Version"),
        },
        judgment,
        layout,
        timing,
        notes: chart_notes,
    })
}
